README.md
# Infrastructure for transactions app
IaC in sh and yml files

## Transactions50 is my final project for Harvards CS50 Web Programing with Python and JavaScript
This app allow users to buy,sell,withdraw,send cryptocurency, to process fiat operations app using Stripe and PayPal API, to do cryprocurency operations app using MetaMask API.

To do this user needs to have account in application, there is three types of accounts: Browsing, Basic and Full. 
 - Browsing - account allowing to check application. Only email and password required
 - Basic - standart account with no KYC, just last name, first name DOB etc . Allows users to deposit, withdraw and send cryptocurency within ETH network.
 - Full - account with KYC verification allows users to deposit fiat,withdraw fiat, buy cryptocurency and sell it.

It consists of 6 microservices:
 - Authentication service 
 - User managment service/API Gateway
 - Payment gateway service
 - Transactions service
 - Walet integration service
 - Notification service
And UI app


![Screenshot 2023-04-23 at 11 20 10 am](https://user-images.githubusercontent.com/71220725/233814165-94bcbd51-b09b-4efd-a63b-c60606e3654c.png)


### Infrastructure: 
All services are containerised appliacitons and all can be bootstrapped using scripts in ```infrastructure``` folder, all additional resources may be bootstraped using docker compose script in ```infrastructure/platform``` it includes:
- Postgres database with credentials
- Postgres database with transactions
- MongoDb with user information
- RedisJSON db with some of users information
- RabbitMQ for comunication beetween services
- Consul client to register microservices with consul
- Consul server to make health requests
- Jaeger server to display Open Telemetry data

## Services description:

### Authentication service 
Django REST API connected to PostgreSQL database with users credentials used to create new user, authenticate user, change user's password or email, delete user's account. As a method of comunication app using HTTP, so its standart REST API.

To secure API http request must include eather API KEY or Authorisation credentials to make successful request.

App running in container and connected to Consul server, Jaeger via OpenTelemtry and postgreSQL server.

* container runs interactivly to deregister service from Consul when it stopped correctly, so it should be stopped only from containers bash *

To run this container running postgres container (```infrastructure/platform/docker-compose.yaml```)(infrastructure repo)  <b>at least</b> required.

All settings of Django app contains in docker run script named ```infrastructure/auth-service-deploy.sh```(infrastructure repo)  and may be specified in it before running, all sensitive data should be stored in Kubernetes or Docker swarm secrets. 

#### To obtain API key:
After starting container, go to containers shell 
        

            docker exec -it <container_id> bash

in there apply migrations

  
            python3 manage.py migrate
   
than go to python shell

 
            python3 manage.py shell
  
in there execute:

    from rest_framework_api_key.models import APIKey

    api_key, key = APIKey.objects.create_key(name="<service_name>")

and print key: 

```print(key)```



### API Gateway / User managment service
Express server built with Typescript using PostgreSQL to store transactions data, MongoDB to store users data and Redis to store frequently used data.
This server will act as gateway from frontend app and will recive all requests from client and then send data to other services via HTTP or RabitMQ.

Routes include : login,register,delete user,edit password, add/edit KYC of level 1 or 2 , get KYC data, get Redis data.

App running in container and connected to Consul server, Jaeger via OpenTelemtry and postgreSQL server.
    
All routes will be secured with JWT authentication with 2h expiration time, to obtain JWT user needs to pass authorisation (login/register in front-end app).

To run this container running postgres container, redis container and mongoDB container (```infrastructure/platform/docker-compose.yaml```)(infrastructure repo)  <b>at least</b> required.

All settings of Express app contains in docker run script named ```infrastructure/userManagment-service-deploy.sh```(infrastructure repo) and ```app/config/*``` in they may be specified in it before running, all sensitive data should be stored in Kubernetes or Docker swarm secrets. 

To do requests to authentication service we need to specify API_KEY in container deployment script (```infrastructure/userManagment-service-deploy.sh```)(infrastructure repo)






CHANGELOG.md
# Changelog

## 0.1.1 (Unreleased)
Docker compose file created for all resources
## 0.1.2 (Unreleased)
Redis container added to docker compose file and Jaeger container added to docker compose file but commented out for now 
## 0.1.0 (Unreleased)
App is in deployment and container run script created

auth-service-deploy.sh
#!/bin/bash
source .env

set -e
CONSUL_HOST=consul-client #change in production
CONSUL_PORT=8500 
JAEGER_HOST="jaeger"
JAEGER_PORT=4317
SERVICE_NAME='Authentication Service'
SERVICE_PORT=8000
DB_NAME='usersPostgres'  #change in production
DB_HOST='users-db-postgres' #change in production
DB_PORT=5432
DB_USER="transactions50adminUsers" #secret
DB_PASSWORD="RabbitMQconfigurationbypullingfromthesystem#Users" #secret
KEY=${DJANGO_KEY}
DJANGO_DEBUG_BOOLEAN=False
SERVICE_HOST="*" #change in production


# Run
docker run \
  --name auth-service \
  --rm \
  -d \
  -e CONSUL_HOST="${CONSUL_HOST}" \
  -e CONSUL_PORT=${CONSUL_PORT} \
  -e JAEGER_HOST=${JAEGER_HOST} \
  -e JAEGER_PORT=${JAEGER_PORT} \
  -e SERVICE_PORT=${SERVICE_PORT} \
  -e SERVICE_NAME="${SERVICE_NAME}" \
  -e SERVICE_HOST="${SERVICE_HOST}" \
  -e KEY="${KEY}" \
  -e DJANGO_DEBUG_BOOLEAN=${DJANGO_DEBUG_BOOLEAN} \
  -e DB_NAME="${DB_NAME}" \
  -e DB_HOST="${DB_HOST}" \
  -e DB_PORT="${DB_PORT}" \
  -e DB_USER="${DB_USER}" \
  -e DB_PASSWORD="${DB_PASSWORD}" \
  -p ${SERVICE_PORT}:8000 \
  --network test \
  authentication-django


#Verify
docker ps


transactions-service-deploy.sh
#!/bin/bash
source .env

set -e
#General configs
NODE_ENV="development" #change in production
PORT=3001

#Postgre configs
POSTGRES_USER="transactions50adminTransactions" #secret
PGHOST="transactions-db-posgres"
PGPORT=5433
PGPASSWORD="RabbitMQconfigurationbypullingfromthesystem#Transactions" #secret


#Consul setting
CONSUL_SERVICE_NAME="Transactions Service"
CONSUL_HOST="consul-client" #change in production
CONSUL_PORT=8500


# RabbitMQ configs
AMQPPORT=5672
AMQPHOST="rabbitmq"

docker run \
  --name transactions-service\
  --rm \
  -it \
  -e NODE_ENV="${NODE_ENV}" \
  -e PORT="${PORT}" \
  -e PGUSER="${POSTGRES_USER}" \
  -e PGHOST="${PGHOST}" \
  -e PGPORT="${PGPORT}" \
  -e PGPASSWORD="${PGPASSWORD}" \
  -e CONSUL_SERVICE_NAME="${CONSUL_SERVICE_NAME}" \
  -e CONSUL_HOST="${CONSUL_HOST}" \
  -e CONSUL_PORT="${CONSUL_PORT}" \
  -e AMQPPORT="${AMQPPORT}" \
  -e AMQPHOST="${AMQPHOST}" \
  -p $PORT:3001 \
  --network test \
  transactions-express


#Verify
docker ps

db-sh-scrips(using-compose-now)/transactions-postgresDB-deploy.sh
#!/bin/bash
set -e
PORT=5433
POSTGRES_USER="transactions50adminTransactions" #secret
POSTGRES_PASSWORD="RabbitMQconfigurationbypullingfromthesystem#Transactions" #secret
DB_NAME='transactions'


# Run
docker run \
  --name transactions-db \
  --rm \
  -d \
  -e POSTGRES_USER="${POSTGRES_USER}" \
  -e PGPORT="${PORT}" \
  -e POSTGRES_PASSWORD="${POSTGRES_PASSWORD}" \
  -p ${PORT}:5432 \
  --network test \
  postgres

#Verify
docker ps


# Copy transactions-schema.sql file to container
docker cp transactions-schema.sql transactions-db:/tmp/
sleep 5

docker exec -it transactions-db   createdb -U ${POSTGRES_USER} ${DB_NAME}
docker exec -it transactions-db psql -U ${POSTGRES_USER} ${DB_NAME} -a -f ./tmp/transactions-schema.sql

# Remove transactions-schema.sql file from container
docker exec -i transactions-db rm /tmp/transactions-schema.sql

userManagment-service-deploy.sh
#!/bin/bash
source .env

set -e
#General configs
NODE_ENV="development" #change in production
PORT=3000
JWT_KEY="development-jwt-key" #secret, change in production(use open ssl)

#Postgre configs
POSTGRES_USER="transactions50adminTransactions" #secret
PGHOST="transactions-db-posgres"
PGPORT=5433
PGPASSWORD="RabbitMQconfigurationbypullingfromthesystem#Transactions" #secret


#Mongodb configs
MONGOSTRING="mongodb://users-db-mongo:27017/users_dev" #secret
MONGOUSER="user" #secret
MONGOPASSWORD="password" #secret

#Redis config
REDIS_STRING="redis://users-redis:6379"


#Consul setting
CONSUL_SERVICE_NAME="Users Managment Service"
CONSUL_HOST="consul-client" #change in production
CONSUL_PORT=8500


# Auth service configurations (django rest app for authentication)
AUTH_API_KEY=${AUTH_API_KEY}  #secret, specify each run (instructions in auth app readme)

# RabbitMQ configs
AMQPPORT=5672
AMQPHOST="rabbitmq"

docker run \
  --name users-managment-service\
  --rm \
  -it \
  -e NODE_ENV="${NODE_ENV}" \
  -e PORT="${PORT}" \
  -e JWT_KEY="${JWT_KEY}" \
  -e PGUSER="${POSTGRES_USER}" \
  -e PGHOST="${PGHOST}" \
  -e PGPORT="${PGPORT}" \
  -e PGPASSWORD="${PGPASSWORD}" \
  -e MONGOSTRING="${MONGOSTRING}" \
  -e MONGOUSER="${MONGOUSER}" \
  -e MONGOPASSWORD="${MONGOPASSWORD}" \
  -e REDIS_STRING="${REDIS_STRING}" \
  -e CONSUL_SERVICE_NAME="${CONSUL_SERVICE_NAME}" \
  -e CONSUL_HOST="${CONSUL_HOST}" \
  -e CONSUL_PORT="${CONSUL_PORT}" \
  -e AUTH_API_KEY="${AUTH_API_KEY}" \
  -e AMQPPORT="${AMQPPORT}" \
  -e AMQPHOST="${AMQPHOST}" \
  -p $PORT:3000 \
  --network test \
  user-managment-express

#Verify
docker ps